### ISSUE TITLE
Action with (missing) optional Dictionary param, ModelBinder maps all request params to Dict

### ISSUE THREAD

**Issue by senritsu** [aspnet/Mvc#3983]
**tldr**
If an optional dictionary parameter is missing in the request, the ModelBinder tries to cram all other parameters into the dictionary. If the parameter is supplied in the request, even as empty or null value, everything works.

**Details**

Given an action like
```
public void SomeAction(string foo, string bar, Dictionary<Guid,string> problematicDict = null) {...}
```
any request with a payload of the form
```
foo = "something"
bar = "something else"
```
results in a `System.FormatException`, as the ModelBinder tries to use `foo` and `bar` as dictionary keys with their respective values for `problematicDict`, but the keys are not valid `Guid` representations.

This can be further illustrated if the dictionary is changed to accept string keys
```
public void SomeAction(string foo, string bar, Dictionary<string,string> problematicDict = null) {...}
```
The same request results in the parameters `foo` and `bar` having the correct values set In addition to that, the dictionary also has a Count of 2, with keys `foo` and `bar` and the values from the request parameters.

**Workaround**

If `problematicDict` is added to the request as a parameter, even with an empty value or `null` as value, everything works just fine. This leads to believe that the default specified in the action is not properly handled.

**Comment by rynowak** [comment]
What values do you intend to go into the dictionary? How do you expect this to work?

It doesn't sound right that `foo` and `bar` are used as keys in this case, but I'm not convinced this is going to do what you want once that is addressed.

**Comment by senritsu** [comment]
The intention is that request parameter `foo` is mapped to action parameter `foo`, request parameter `bar` is mapped to action parameter `bar`, and if it exists in the request, request parameter `problematicDict` is mapped to action parameter `problematicDict`. If the request parameter is missing, the default of `null` should be used.

**Comment by rynowak** [comment]
You can get this behavior today by using `[ModelBinder(Name="problematicDict")]`.

I agree that the behavior you're seeing is surprising.

**Comment by senritsu** [comment]
I completely forgot about the `[ModelBinder]` attribute. I will try to see if that is enough for our use case.

The behaviour exhibited seems pretty unintentional nevertheless.

**Comment by senritsu** [comment]
When manually specifying the binding name as described by @rynowak, it does indeed bind correctly. Only the regular parameters are mapped, and the dictionary uses the default of `null`.

**Comment by Eilon** [comment]
Moving to Backlog.

**Comment by senritsu** [comment]
Since a colleague just stumbled across that issue again, and i re-discovered this issue while searching for a solution, has anything happened with this issue or is there an ETA?

**Comment by Eilon** [comment]
Moving out of backlog for consideration.

**Comment by mkArtakMSFT** [comment]
Moved this back to the `Backlog` as this is not a priority at the moment. Neither there was much community involvement with this ask.
